engine/notifications.js

```diff
@@ -0,0 +1,283 @@
+import redis from '../engine/redis-helper.js';
+import webPush from 'web-push';
+
+const ttl = 2419200;
+
+if (!process.env.GCM_API_KEY) {
+  console.warn('Set the GCM_API_KEY environment variable to support GCM');
+}
+webPush.setGCMAPIKey(process.env.GCM_API_KEY);
+
+// check if such device exists in database, reject if not
+function checkDeviceId(deviceId) {
+  return redis.hgetall(`device-${deviceId}`)
+  .then(device => {
+    if (!device || !device.endpoint) {
+      throw new Error('Not Found');
+    }
+  });
+}
+
+// get all registrations for a deviceId
+function getRegisteredFeatures(deviceId) {
+  return checkDeviceId(deviceId)
+  .then(() => redis.smembers(`${deviceId}-notifications`));
+}
+
+function sendWebPush(endpoint, key, auth, deviceId, payload) {
+  payload = JSON.stringify(payload);
+  if (!key || !auth) {
+    return redis.set(`${deviceId}-payload`, payload)
+    .then(() =>
+      webPush.sendNotification(endpoint, { TTL: ttl })
+    );
+  }
+  return webPush.sendNotification(endpoint, {
+    TTL: ttl,
+    userPublicKey: key,
+    userAuth: auth,
+    payload,
+  });
+}
+
+/*
+ * send confirmation even if deviceId is already deleted
+ */
+function sendConfirmation(endpoint, key, auth, deviceId, features) {
+  const payload = {
+    title: 'Registration changed',
+    body: 'You\'re not registered to any feature',
+  };
+  if (features && features.length > 0) {
+    const numberOfFeatures = features.length;
+    const message = 'Registered to';
+    if (features.indexOf('all') >= 0) {
```

Comment:
There is a slight chance of having `features == [null, 'all']` - I've seen it once or twice on "living" site, but I'm unable to reproduce.